httpx/client.py

```diff
@@ -476,7 +476,12 @@ def setup_auth(
 
             if not allow_redirects:
                 response.call_next = functools.partial(
-                    self.send_handling_redirects,
+                    # NOTE: not using 'self.send_handling_redirects' here, because
+                    # 'call_next' must reference a function (instead
+                    # of a method). This ensures that 'inspect.iscoroutinefunction()'
+                    # checks behave properly.
+                    Client.send_handling_redirects,
+                    self,
```

Comment:
Yes, that's surprising, and essentially boils down to…

```python
import inspect

class Foo:
    async def bar(self):
        pass

foo = Foo()
print(inspect.iscoroutinefunction(foo.bar))  # False
```

`functools.partial` is not in fault here, since `inspect.is*function` checks unwrap any `partial` down to the underlying function/method.

The "issue" is that `inspect.is*function` has a check on `inspect.isfunction()` (which, er, seems logicial because they both have "function" in there). But in this case `foo.bar` refers to a *method*, [which is different from a function](https://stackoverflow.com/questions/17019949/why-is-there-a-difference-between-inspect-ismethod-and-inspect-isfunction-from-p) in Python 3.

To get a function, we must reach to the unbound method, so:

```python
print(inspect.iscoroutinefunction(Foo.bar))  # True
```

Reply:
Strangely, this workaround doesn't work on 3.6 and 3.7 — CI is failing for those versions. 😕